tests/test_aws.py

```diff
@@ -371,3 +377,153 @@ def test_stream_override(self):
         stream = log_output.get_handler()
         self.assertTrue(stream.log_stream == 'testing')
+
+
+def test_url_socket_retry(monkeypatch):
+    monkeypatch.setattr(time, "sleep", lambda x: x)
+
+    # case for unknown error
+    fvalues = [URLError(socket.error(104, 'Connection reset by peer')),
+               URLError(socket.gaierror(8, 'Name or node unknown'))]
+
+    def freturns():
+        ret = fvalues.pop()
+        if isinstance(ret, URLError):
+            raise ret
+        return ret
+
+    with pytest.raises(URLError) as ecm:
+        aws.url_socket_retry(freturns)
+
+    assert 'Name or node unknown' in str(ecm.value)
+
+    # case for retry exhaustion
+    fvalues[:] = [
+        URLError(socket.error(110, 'Connection timed out')),
+        URLError(socket.error(110, 'Connection timed out')),
+        URLError(socket.error(110, 'Connection timed out')),
+        URLError(socket.error(110, 'Connection timed out')),
+    ]
+
+    with pytest.raises(URLError) as ecm:
+        aws.url_socket_retry(freturns)
+
+    # case for success
+    fvalues[:] = [
+        URLError(socket.error(110, 'Connection timed out')),
+        42
+    ]
+
+@pytest.mark.skipif(sys.version_info < (3, 9), reason="requires python3.9 or higher")
+def test_http_socket_retry(monkeypatch):
+    monkeypatch.setattr(time, "sleep", lambda x: x)
+
+    # case for unknown error
+    fvalues = [HTTPError('https://lwn.net', 404, 'Unknown', {}, None)]
+
+    def freturns():
+        ret = fvalues.pop()
+        if isinstance(ret, URLError):
+            raise ret
+        return ret
+
+    with pytest.raises(URLError) as ecm:
+        aws.url_socket_retry(freturns)
+
+    assert 'Unknown' in str(ecm.value)
+
+    # case for retry exhaustion    
+    fvalues[:] = [
+        HTTPError('https://lwn.net', 503, 'Slow Down', {}, None),
+        HTTPError('https://lwn.net', 503, 'Slow Down', {}, None),
+        HTTPError('https://lwn.net', 503, 'Slow Down', {}, None),
+        HTTPError('https://lwn.net', 503, 'Slow Down', {}, None),
+    ]
+
+    # this is really an indirect assertion that is ascertained via
+    # coverage.
+    with pytest.raises(URLError) as ecm:
+        aws.url_socket_retry(freturns)    
+
+    # case for success
+    fvalues[:] = [
+        HTTPError('https://lwn.net', 503, 'Slow Down', {}, None),
+        42
+    ]
+
+    assert aws.url_socket_retry(freturns) == 42
+
+
+def test_default_bucket_region_with_no_s3():
+    output_dir = "/tmp"
+    conf = Config.empty(output_dir=output_dir)
+    aws._default_bucket_region(conf)
+    assert output_dir == conf.output_dir
+
+
+def test_default_bucket_region_with_explicit_region():
+    output_dir = "s3://aws?region=xyz"
+    conf = Config.empty(output_dir=output_dir)
+    aws._default_bucket_region(conf)
+    assert output_dir == conf.output_dir
+
+
+@vcr.use_cassette(
+    'tests/data/vcr_cassettes/test_output/default_bucket_region_public.yaml', record_mode='all')
```

Comment:
mode here should be none to use recording afaics

Reply:
actually we should probably remove all of the record_mode as the default is adequate for offline tests (aka once).